frame/democracy/src/lib.rs

```diff
@@ -827,9 +922,24 @@ impl<T: Trait> Module<T> {
 	}
 
 	/// Enact a proposal from a referendum.
-	fn enact_proposal(proposal: T::Proposal, index: ReferendumIndex) {
-		let ok = proposal.dispatch(system::RawOrigin::Root.into()).is_ok();
-		Self::deposit_event(RawEvent::Executed(index, ok));
+	fn enact_proposal(proposal_hash: T::Hash, index: ReferendumIndex) -> Result {
+		if let Some((encoded_proposal, who, amount, _)) = <Preimages<T>>::take(&proposal_hash) {
+			if let Ok(proposal) = T::Proposal::decode(&mut &encoded_proposal[..]) {
```

Comment:
It seems there is a way that a reserved balance is permanently stuck.

If the `encoded_proposal` fails to decode, you will never get to the `unreserve` action below.

Since you `take` on the `Preimages`, you will also not be able to `reap_preimage`.

Seems like the unreserve step should happen before the decode or in both of the else statements.

Reply:
if you propose a hash that resolves to a blob which is not a valid `Proposal` encoding, then you deserve to lose access to the reserved funds.

Reply:
addressed.